In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv('mushrooms.csv')
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [3]:
df.shape

(8124, 23)

The dataset consists of 8124 rows and 23 columns

In [4]:
df.dtypes

class                       object
cap-shape                   object
cap-surface                 object
cap-color                   object
bruises                     object
odor                        object
gill-attachment             object
gill-spacing                object
gill-size                   object
gill-color                  object
stalk-shape                 object
stalk-root                  object
stalk-surface-above-ring    object
stalk-surface-below-ring    object
stalk-color-above-ring      object
stalk-color-below-ring      object
veil-type                   object
veil-color                  object
ring-number                 object
ring-type                   object
spore-print-color           object
population                  object
habitat                     object
dtype: object

All columns in the dataset are of categorical type

### Preprocessing 

In [5]:
df.isnull().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

There are no null values in the dataset

In [6]:
cols=df.columns
for i in cols:
    print(i)
    print(df[i].unique(),'\n')

class
['p' 'e'] 

cap-shape
['x' 'b' 's' 'f' 'k' 'c'] 

cap-surface
['s' 'y' 'f' 'g'] 

cap-color
['n' 'y' 'w' 'g' 'e' 'p' 'b' 'u' 'c' 'r'] 

bruises
['t' 'f'] 

odor
['p' 'a' 'l' 'n' 'f' 'c' 'y' 's' 'm'] 

gill-attachment
['f' 'a'] 

gill-spacing
['c' 'w'] 

gill-size
['n' 'b'] 

gill-color
['k' 'n' 'g' 'p' 'w' 'h' 'u' 'e' 'b' 'r' 'y' 'o'] 

stalk-shape
['e' 't'] 

stalk-root
['e' 'c' 'b' 'r' '?'] 

stalk-surface-above-ring
['s' 'f' 'k' 'y'] 

stalk-surface-below-ring
['s' 'f' 'y' 'k'] 

stalk-color-above-ring
['w' 'g' 'p' 'n' 'b' 'e' 'o' 'c' 'y'] 

stalk-color-below-ring
['w' 'p' 'g' 'b' 'n' 'e' 'y' 'o' 'c'] 

veil-type
['p'] 

veil-color
['w' 'n' 'o' 'y'] 

ring-number
['o' 't' 'n'] 

ring-type
['p' 'e' 'l' 'f' 'n'] 

spore-print-color
['k' 'n' 'u' 'h' 'w' 'r' 'o' 'y' 'b'] 

population
['s' 'n' 'a' 'v' 'y' 'c'] 

habitat
['u' 'g' 'm' 'd' 'p' 'w' 'l'] 



Under the stalk-root column we can see one of the unique values as '?' which represents missing values

In [7]:
df['stalk-root'].value_counts()

b    3776
?    2480
e    1120
c     556
r     192
Name: stalk-root, dtype: int64

There are 2480 missing values under the column which needs to be filled

In [8]:
df['stalk-root'].replace('?', np.NaN, inplace=True)
df['stalk-root'].isnull().mean()

0.3052683407188577

In [9]:
df['stalk-root'].isnull().sum()

2480

In [10]:
df['stalk-root']=df['stalk-root'].fillna(df['stalk-root'].mode()[0])
df['stalk-root'].value_counts()

b    6256
e    1120
c     556
r     192
Name: stalk-root, dtype: int64

In [11]:
df.drop(columns=['stalk-color-above-ring','stalk-color-below-ring','veil-type'],axis=1,inplace=True)
df.shape

(8124, 20)

In [12]:
df['class']=df['class'].map({'p':0,'e':1})
df['class'].value_counts()

1    4208
0    3916
Name: class, dtype: int64

In [13]:
X=df.drop('class',axis=1)
y=df['class']

### Feature encoding

In [14]:
X=pd.get_dummies(X,drop_first=True)
pd.pandas.set_option('display.max_columns',None)
X.head()

cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  cap-shape_x  \
0            0            0            0            0            1   
1            0            0            0            0            1   
2            0            0            0            0            0   
3            0            0            0            0            1   
4            0            0            0            0            1   

   cap-surface_g  cap-surface_s  cap-surface_y  cap-color_c  cap-color_e  \
0              0              1              0            0            0   
1              0              1              0            0            0   
2              0              1              0            0            0   
3              0              0              1            0            0   
4              0              1              0            0            0   

   cap-color_g  cap-color_n  cap-color_p  cap-color_r  cap-color_u  \
0            0            1            0            0            0   
1            0            0            0            0            0   
2            0            0            0            0            0   
3            0            0            0            0            0   
4            1            0            0            0            0   

   cap-color_w  cap-color_y  bruises_t  odor_c  odor_f  odor_l  odor_m  \
0            0            0          1       0       0       0       0   
1            0            1          1       0       0       0       0   
2            1            0          1       0       0       1       0   
3            1            0          1       0       0       0       0   
4            0            0          0       0       0       0       0   

   odor_n  odor_p  odor_s  odor_y  gill-attachment_f  gill-spacing_w  \
0       0       1       0       0                  1               0   
1       0       0       0       0                  1               0   
2       0       0       0       0                  1               0   
3       0       1       0       0                  1               0   
4       1       0       0       0                  1               1   

   gill-size_n  gill-color_e  gill-color_g  gill-color_h  gill-color_k  \
0            1             0             0             0             1   
1            0             0             0             0             1   
2            0             0             0             0             0   
3            1             0             0             0             0   
4            0             0             0             0             1   

   gill-color_n  gill-color_o  gill-color_p  gill-color_r  gill-color_u  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             1             0             0             0             0   
3             1             0             0             0             0   
4             0             0             0             0             0   

   gill-color_w  gill-color_y  stalk-shape_t  stalk-root_c  stalk-root_e  \
0             0             0              0             0             1   
1             0             0              0             1             0   
2             0             0              0             1             0   
3             0             0              0             0             1   
4             0             0              1             0             1   

   stalk-root_r  stalk-surface-above-ring_k  stalk-surface-above-ring_s  \
0             0                           0                           1   
1             0                           0                           1   
2             0                           0                           1   
3             0                           0                           1   
4             0                           0                           1   

   stalk-surface-above

In [15]:
X.shape

(8124, 78)

### Train test splitting

In [16]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=42,test_size=.3,stratify = y) #30% data is for testing
print(y_train.value_counts())
print(y_test.value_counts())

1    2945
0    2741
Name: class, dtype: int64
1    1263
0    1175
Name: class, dtype: int64


### Modelling

In [17]:

from sklearn.naive_bayes import GaussianNB,MultinomialNB,BernoulliNB,CategoricalNB
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

In [18]:

models = {
    '     Gaussian Naive Bayes' : GaussianNB(),
    '  Multinomial Naive Bayes' : MultinomialNB(),
    '    Bernoulli Naive Bayes' : BernoulliNB(),
    '  Categorical Naive Bayes' : CategoricalNB(),
    '      Logistic Regression' : LogisticRegression(),
    '     KNeighborsClassifier' : KNeighborsClassifier(),
}

In [19]:

for name, model in models.items():
    model.fit(X_train, y_train)
    print(name + ' is trained.')

     Gaussian Naive Bayes is trained.
  Multinomial Naive Bayes is trained.
    Bernoulli Naive Bayes is trained.
  Categorical Naive Bayes is trained.
      Logistic Regression is trained.
     KNeighborsClassifier is trained.


In [20]:

def get_accuracy_score(y_test,prediction):
    accuracy = accuracy_score(y_test,prediction)
    return accuracy
def get_precision_score(y_test,prediction):
    precision = precision_score(y_test,prediction)
    return precision
def get_recall_score(y_test,prediction):
    recall = recall_score(y_test,prediction)
    return recall
def get_f1_score(y_test,prediction):
    f1 = f1_score(y_test,prediction)
    return f1
def get_cross_val_score(model,X_train,y_train):
    cross_val_scores = cross_val_score(model,X_train,y_train,cv=10)
    avg_cross_val_score=cross_val_scores.mean()
    return avg_cross_val_score
def get_roc_auc_score(y_test,prediction):
    roc_auc = roc_auc_score(y_test,prediction)
    return roc_auc

In [21]:
met= pd.DataFrame(columns = [ "Accuracy", "F1_score","Recall","Precision", "Cross_validation","ROC_AUC"],
                  index=['     Gaussian Naive Bayes','  Multinomial Naive Bayes','    Bernoulli Naive Bayes',
                        '  Categorical Naive Bayes','      Logistic Regression','     KNeighborsClassifier'])

In [22]:
for name, model in models.items():
    prediction = model.predict(X_test)
    acc = get_accuracy_score(y_test, prediction)
    met.at[name,"Accuracy"]=acc
    f1 = get_f1_score(y_test, prediction)
    met.at[name,"F1_score"]=f1
    recal = get_recall_score(y_test, prediction)
    met.at[name,"Recall"]=recal
    prec = get_precision_score(y_test, prediction)
    met.at[name,"Precision"]=prec
    xval = get_cross_val_score(model,X_train,y_train)
    met.at[name,"Cross_validation"]=xval
    roc = get_roc_auc_score(y_test, prediction)
    met.at[name,"ROC_AUC"]=roc
met   

Accuracy  F1_score    Recall Precision  \
     Gaussian Naive Bayes  0.996719  0.996825  0.994458  0.999204   
  Multinomial Naive Bayes  0.965546  0.967816       1.0  0.937639   
    Bernoulli Naive Bayes  0.946678  0.950795  0.994458  0.910805   
  Categorical Naive Bayes  0.946678  0.950795  0.994458  0.910805   
      Logistic Regression   0.99959  0.999604       1.0  0.999209   
     KNeighborsClassifier       1.0       1.0       1.0       1.0   

                          Cross_validation   ROC_AUC  
     Gaussian Naive Bayes         0.994195  0.996803  
  Multinomial Naive Bayes         0.957088  0.964255  
    Bernoulli Naive Bayes         0.944248  0.944888  
  Categorical Naive Bayes         0.944248  0.944888  
      Logistic Regression         0.999648  0.999574  
     KNeighborsClassifier              1.0       1.0